### PR TITLE
Add CoProduct and Either

### PR DESCRIPTION
**Pull request by mstn**
CoProduct and "proof" that Idris Either is a CoProduct in Idris Category.

### CODE REVIEW

File: src/Idris/EitherAsCoProduct.lidr [hunk 1]
```diff
@@ -0,0 +1,56 @@
+> module Idris.CoProduct
```
Comment (marcosh) [inline src/Idris/EitherAsCoProduct.lidr]: could you please add the license on top of the files?

Comment (marcosh) [inline src/Idris/EitherAsCoProduct.lidr]: Could you call this module with the name of the file?

File: src/Idris/EitherAsCoProduct.lidr [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+> import Basic.Category
+> import Limits.CoProduct
+>
+> applyLeftOrRight :
+>      (a, b, c : Type)
+>   -> (f : ExtensionalTypeMorphism a c)
+>   -> (g : ExtensionalTypeMorphism b c)
+>   -> (ExtensionalTypeMorphism (Either a b) c)
+> applyLeftOrRight a b c f g = MkExtensionalTypeMorphism
+>   (\x => case x of
```
Comment (marcosh) [inline src/Idris/EitherAsCoProduct.lidr]: you could write this function as `either (func f) (func g)`

File: src/Idris/EitherAsCoProduct.lidr [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+>           Left y => func f y
+>           Right y => func g y
+>   )
+>
+> cfgLeftCompose :
```
Comment (marcosh) [inline src/Idris/EitherAsCoProduct.lidr]: what does `cfg` stand for? Could you find a more expressive name, maybe?

File: src/Idris/EitherAsCoProduct.lidr [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+>   -> (f : ExtensionalTypeMorphism a c)
+>   -> (g : ExtensionalTypeMorphism b c)
+>   -> (h : ExtensionalTypeMorphism (Either a b) c)
+>   -> (left  : f = extCompose a (Either a b) c (MkExtensionalTypeMorphism Left) h)
+>   -> (right : g = extCompose b (Either a b) c (MkExtensionalTypeMorphism Right) h)
+>   -> applyLeftOrRight a b c f g = h
+> cfgUnique a b c (MkExtensionalTypeMorphism f) (MkExtensionalTypeMorphism g) (MkExtensionalTypeMorphism h) left right =
+>   funExt(
+>     \x => case x of
+>          Left y => cong {f =(\extFun => apply (func extFun) y)} left
```
Comment (marcosh) [inline src/Idris/EitherAsCoProduct.lidr]: could you keep 2 spaces of indentation?

File: src/Idris/EitherAsCoProduct.lidr [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+>          Right y => cong {f =(\extFun => apply (func extFun) y)} right
+>   )
+>
+> eitherToCoProduct : (a, b : Type) -> CoProduct Idris.TypesAsCategoryExtensional.typesAsCategoryExtensional a b
+> eitherToCoProduct a b = MkCoProduct
+>   ( Either a b )
```
Comment (marcosh) [inline src/Idris/EitherAsCoProduct.lidr]: do not add whitespace after `(` and before `)`

File: src/Limits/CoProduct.lidr [hunk 6]
```diff
@@ -0,0 +1,35 @@
+> module Limits.CoProduct
```
Comment (marcosh) [inline src/Limits/CoProduct.lidr]: please move to the `Colimits` folder

File: src/Limits/CoProduct.lidr [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+>
+> import Basic.Category
+>
+> %access public export
+> %default total
+>
+> record CoProduct (cat : Category) (a : obj cat) (b : obj cat) where
+>   constructor MkCoProduct
+>   payload: obj cat
```
Comment (marcosh) [inline src/Limits/CoProduct.lidr]: please call this `carrier`

File: src/Limits/CoProduct.lidr [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+>   inl: mor cat a payload
+>   inr: mor cat b payload
+>   cfgMorphism:
```
Comment (marcosh) [inline src/Limits/CoProduct.lidr]: for the `Product` definition we are working on, we defined a type like

```
record CommutingMorphism (cat : Category)
                         (x : obj cat) (a : obj cat) (b : obj cat) (carrier : obj cat)
                         (pi1 : mor cat carrier a) (pi2 : mor cat carrier b)
                         (f : mor cat x a) (g : mor cat x b) where
  constructor MkCommutingMorphism
  challenger         : mor cat x carrier
  commutativityLeft  : compose cat x carrier a challenger pi1 = f
  commutativityRight : compose cat x carrier b challenger pi2 = g
```

to describe morphisms which commutes with the projections. What do you think? Could it be a good idea to use a similar approach also here?

File: src/Idris/EitherAsCoProduct.lidr [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+>
+> cfgUnique :
+>      (a, b, c : Type)
+>   -> (f : ExtensionalTypeMorphism a c)
+>   -> (g : ExtensionalTypeMorphism b c)
+>   -> (h : ExtensionalTypeMorphism (Either a b) c)
+>   -> (left  : f = extCompose a (Either a b) c (MkExtensionalTypeMorphism Left) h)
+>   -> (right : g = extCompose b (Either a b) c (MkExtensionalTypeMorphism Right) h)
+>   -> applyLeftOrRight a b c f g = h
+> cfgUnique a b c (MkExtensionalTypeMorphism f) (MkExtensionalTypeMorphism g) (MkExtensionalTypeMorphism h) left right =
```
Comment (marcosh) [inline src/Idris/EitherAsCoProduct.lidr]: there is no need to expand `f` and `g` here

File: src/Limits/CoProduct.lidr (hunk 8 above)
Comment (mstn) [reply src/Limits/CoProduct.lidr]: Definitely more readable.

**Comment by mstn** [comment]
Thanks for feedback. It wasn't too picky. Code readability improved.

I am going to work on cocones next. What do you think? I do not have a plan. I am just following classical textbooks, just to practice a bit.

**Comment by marcosh** [comment]
@mstn defining cones and cocones could be nice and fun, but we are trying to avoid the general definition of limits and colimits for now, because they bring in complexity which we don't want to manage for the moment. Still, if you want to take a shot a them, that could save us some work!

**Pull request merged**